#Step:1-

## Demonstration of text being converted to tokens.
  

In [ ]:
test_string = '''
Artificial Intelligence (AI) is changing the world.
人工知能は世界を変えています。

def greet(name):
    print(f"Hello, {name}!")  # Greet the user

# こんにちは、ユーザーさん！
greet("世界")

Data analysis and natural language processing (NLP) are core parts of AI.
データ分析と言語処理はAIの中核です。
'''

tokens = test_string.encode("utf-8") # raw bytes
tokens = list(map(int, tokens)) # convert to a list of integers in range 0..255 for convenience
print('---')
print(test_string)
print("length:", len(test_string))
print('---')
print(tokens)
print("length:", len(tokens))

---

Artificial Intelligence (AI) is changing the world.
人工知能は世界を変えています。

def greet(name):
    print(f"Hello, {name}!")  # Greet the user

# こんにちは、ユーザーさん！
greet("世界")

Data analysis and natural language processing (NLP) are core parts of AI.
データ分析と言語処理はAIの中核です。

length: 258
---
[10, 65, 114, 116, 105, 102, 105, 99, 105, 97, 108, 32, 73, 110, 116, 101, 108, 108, 105, 103, 101, 110, 99, 101, 32, 40, 65, 73, 41, 32, 105, 115, 32, 99, 104, 97, 110, 103, 105, 110, 103, 32, 116, 104, 101, 32, 119, 111, 114, 108, 100, 46, 10, 228, 186, 186, 229, 183, 165, 231, 159, 165, 232, 131, 189, 227, 129, 175, 228, 184, 150, 231, 149, 140, 227, 130, 146, 229, 164, 137, 227, 129, 136, 227, 129, 166, 227, 129, 132, 227, 129, 190, 227, 129, 153, 227, 128, 130, 10, 10, 100, 101, 102, 32, 103, 114, 101, 101, 116, 40, 110, 97, 109, 101, 41, 58, 10, 32, 32, 32, 32, 112, 114, 105, 110, 116, 40, 102, 34, 72, 101, 108, 108, 111, 44, 32, 123, 110, 97, 109, 101, 125, 33, 34, 41, 32, 32, 35, 32, 71, 114, 101, 101, 1

### The reason that encoding bytes are more (352>258) because it follows:

UTF-8 is a **variable-length encoding**, meaning:

- ASCII characters (e.g., `A-Z, 0-9, !@#`) → **1 byte**
- Latin and extended characters (e.g., `é, ñ, ü`) → **2 bytes**
- Non-Latin characters (e.g., `漢字, अ, У`) → **3 or 4 bytes**
- Emojis (e.g., `😄, 🇺🇳`) → **4 bytes)**

# Step 2:

In [ ]:
# Funtion to colab
def get_stats(ids):
    counts = {}
    for pair in zip(ids, ids[1:]): # Pythonic way to iterate consecutive elements
        counts[pair] = counts.get(pair, 0) + 1
    return counts

stats = get_stats(tokens)
print(sorted(((v,k) for k,v in stats.items()), reverse=True)) #This doing nothing but reversing the

[(16, (227, 129)), (6, (101, 32)), (5, (227, 131)), (5, (227, 130)), (5, (114, 101)), (4, (110, 103)), (4, (110, 97)), (4, (97, 110)), (4, (32, 32)), (3, (228, 184)), (3, (227, 128)), (3, (188, 227)), (3, (131, 188)), (3, (129, 175)), (3, (116, 40)), (3, (115, 32)), (3, (105, 110)), (3, (101, 116)), (3, (101, 101)), (3, (97, 108)), (3, (65, 73)), (3, (41, 32)), (3, (32, 112)), (3, (32, 97)), (3, (10, 10)), (2, (231, 149)), (2, (184, 150)), (2, (166, 227)), (2, (153, 227)), (2, (150, 231)), (2, (149, 140)), (2, (147, 227)), (2, (130, 147)), (2, (130, 10)), (2, (129, 153)), (2, (128, 130)), (2, (116, 104)), (2, (115, 105)), (2, (114, 116)), (2, (112, 114)), (2, (111, 114)), (2, (110, 116)), (2, (109, 101)), (2, (108, 108)), (2, (108, 32)), (2, (105, 115)), (2, (104, 101)), (2, (103, 114)), (2, (103, 101)), (2, (103, 32)), (2, (102, 32)), (2, (101, 108)), (2, (99, 101)), (2, (97, 116)), (2, (97, 114)), (2, (97, 109)), (2, (46, 10)), (2, (35, 32)), (2, (34, 41)), (2, (32, 116)), (2, (32, 9

## Step 3:
What we are doing here:
  1.Finding the pair of consequitive token.
  2. Merging these consequitive token into a unicode token i furtherance of token dictionary so far formeed that means merging and assigning this pair to a new token ID in continues fashion .

In [ ]:
def get_stats(ids):
    counts = {}
    for pair in zip(ids, ids[1:]): # Pythonic way to iterate consecutive elements
        counts[pair] = counts.get(pair, 0) + 1
    return counts

stats = get_stats(tokens)
# print(stats)
# print(sorted(((v,k) for k,v in stats.items()), reverse=True))

In [ ]:
top_pair = max(stats, key=stats.get)
top_pair

(227, 129)

In [ ]:
def merge(ids, pair, idx):
  # in the list of ints (ids), replace all consecutive occurences of pair with the new token idx
  newids = []
  i = 0
  while i < len(ids):
    # if we are not at the very last position AND the pair matches, replace it
    if i < len(ids) - 1 and ids[i] == pair[0] and ids[i+1] == pair[1]:
      newids.append(idx)
      i += 2
    else:
      newids.append(ids[i])
      i += 1
  return newids

tokens2 = merge(tokens, top_pair, 256) # Here we are making a fresh updated set of tokens .
print(tokens2)
print("length:", len(tokens2))

[10, 65, 114, 116, 105, 102, 105, 99, 105, 97, 108, 32, 73, 110, 116, 101, 108, 108, 105, 103, 101, 110, 99, 101, 32, 40, 65, 73, 41, 32, 105, 115, 32, 99, 104, 97, 110, 103, 105, 110, 103, 32, 116, 104, 101, 32, 119, 111, 114, 108, 100, 46, 10, 228, 186, 186, 229, 183, 165, 231, 159, 165, 232, 131, 189, 256, 175, 228, 184, 150, 231, 149, 140, 227, 130, 146, 229, 164, 137, 256, 136, 256, 166, 256, 132, 256, 190, 256, 153, 227, 128, 130, 10, 10, 100, 101, 102, 32, 103, 114, 101, 101, 116, 40, 110, 97, 109, 101, 41, 58, 10, 32, 32, 32, 32, 112, 114, 105, 110, 116, 40, 102, 34, 72, 101, 108, 108, 111, 44, 32, 123, 110, 97, 109, 101, 125, 33, 34, 41, 32, 32, 35, 32, 71, 114, 101, 101, 116, 32, 116, 104, 101, 32, 117, 115, 101, 114, 10, 10, 35, 32, 256, 147, 227, 130, 147, 256, 171, 256, 161, 256, 175, 227, 128, 129, 227, 131, 166, 227, 131, 188, 227, 130, 182, 227, 131, 188, 256, 149, 227, 130, 147, 239, 188, 129, 10, 103, 114, 101, 101, 116, 40, 34, 228, 184, 150, 231, 149, 140, 34, 41, 1

# Step:4
What we are doing is:
  1. looping the merge
  2. demonstrating each merge till no merge can follow.

In [ ]:
vocab_size = 276 # the desired final vocabulary size
num_merges = vocab_size - 256
ids = list(tokens) # copy so we don't destroy the original list

merges = {} # (int, int) -> int
for i in range(num_merges):
  stats = get_stats(ids)
  pair = max(stats, key=stats.get)
  idx = 256 + i
  print(f"merging {pair} into a new token {idx}")
  ids = merge(ids, pair, idx)
  merges[pair] = idx

merging (227, 129) into a new token 256
merging (101, 32) into a new token 257
merging (227, 130) into a new token 258
merging (227, 131) into a new token 259
merging (97, 110) into a new token 260
merging (32, 32) into a new token 261
merging (97, 108) into a new token 262
merging (65, 73) into a new token 263
merging (115, 32) into a new token 264
merging (105, 110) into a new token 265
merging (256, 175) into a new token 266
merging (228, 184) into a new token 267
merging (227, 128) into a new token 268
merging (10, 10) into a new token 269
merging (114, 101) into a new token 270
merging (270, 101) into a new token 271
merging (271, 116) into a new token 272
merging (110, 97) into a new token 273
merging (259, 188) into a new token 274
merging (114, 116) into a new token 275


# Step 5:
  1. Finaly we are printing number of token.
  2. printing what was the lenght before merging.
  3. A Compression Factor which is comprassion ratio which will bsically help us judge our merging process was usefull till which level.

In [ ]:
print("tokens length:", len(tokens))
print("ids length:", len(ids))
print(f"compression ratio: {len(tokens) / len(ids):.2f}X")

tokens length: 352
ids length: 272
compression ratio: 1.29X


# Building the whole cycle of encoding and decoding .

## Designing our own Encoder funtion which going to do multiple murges.

In [ ]:
def encode(text):
  # given a string, return list of integers (the tokens)
  tokens = list(text.encode("utf-8"))
  while len(tokens) >= 2:
    stats = get_stats(tokens)
    pair = min(stats, key=lambda p: merges.get(p, float("inf")))
    if pair not in merges:
      break # nothing else can be merged
    idx = merges[pair]
    tokens = merge(tokens, pair, idx)
  return tokens

print(encode(test_string))

[10, 65, 275, 105, 102, 105, 99, 105, 262, 32, 73, 110, 116, 101, 108, 108, 105, 103, 101, 110, 99, 257, 40, 263, 41, 32, 105, 264, 99, 104, 260, 103, 265, 103, 32, 116, 104, 257, 119, 111, 114, 108, 100, 46, 10, 228, 186, 186, 229, 183, 165, 231, 159, 165, 232, 131, 189, 266, 267, 150, 231, 149, 140, 258, 146, 229, 164, 137, 256, 136, 256, 166, 256, 132, 256, 190, 256, 153, 268, 130, 269, 100, 101, 102, 32, 103, 272, 40, 273, 109, 101, 41, 58, 10, 261, 261, 112, 114, 265, 116, 40, 102, 34, 72, 101, 108, 108, 111, 44, 32, 123, 273, 109, 101, 125, 33, 34, 41, 261, 35, 32, 71, 272, 32, 116, 104, 257, 117, 115, 101, 114, 269, 35, 32, 256, 147, 258, 147, 256, 171, 256, 161, 266, 268, 129, 259, 166, 274, 258, 182, 274, 256, 149, 258, 147, 239, 188, 129, 10, 103, 272, 40, 34, 267, 150, 231, 149, 140, 34, 41, 269, 68, 97, 116, 97, 32, 260, 262, 121, 115, 105, 264, 260, 100, 32, 273, 116, 117, 114, 262, 32, 108, 260, 103, 117, 97, 103, 257, 112, 114, 111, 99, 101, 115, 115, 265, 103, 32, 40, 7

In [ ]:
size=encode(test_string)
print(len(size))

272


In [ ]:
print("tokens length:", len(tokens))
print("ids length:", len(ids))
print(f"compression ratio: {len(tokens) / len(ids):.2f}X")

tokens length: 352
ids length: 272
compression ratio: 1.29X


## Building Decoder to get back the original sentance.

In [ ]:
vocab = {idx: bytes([idx]) for idx in range(256)}
for (p0, p1), idx in merges.items():
    vocab[idx] = vocab[p0] + vocab[p1]

def decode(ids):
  # given ids (list of integers), return Python string
  tokens = b"".join(vocab[idx] for idx in ids)
  text = tokens.decode("utf-8", errors="replace")
  return text

In [ ]:
print(decode(encode(test_string)))


Artificial Intelligence (AI) is changing the world.
人工知能は世界を変えています。

def greet(name):
    print(f"Hello, {name}!")  # Greet the user

# こんにちは、ユーザーさん！
greet("世界")

Data analysis and natural language processing (NLP) are core parts of AI.
データ分析と言語処理はAIの中核です。

